### PR TITLE
[Hub Generated] Review request for Microsoft.Kusto to add version stable/2022-02-01

### PR DESCRIPTION
**Pull request by docohe**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/docohe/azure-rest-api-specs/dev-azure-kusto-Microsoft.Kusto-2022-02-01...main/).

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [x] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by msyyc** [comment]
According to log info, there is unproper definition:
```
fatal   | Process() cancelled due to failure
error   |   Error: $ref to original location 'file:///_/azure-rest-api-specs/specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2022-02-01/kusto.json#/parameters/SubscriptionIdParameter' is not found in the new refs collection
```

**Comment by docohe** [comment]
@msyyc The SubscriptionIdParameter is defined in the swagger in line 6102. What am I missing? It was not change in the version.
The only thing that was changed that I added x-ms-path, maybe it is because of that?

**Comment by msyyc** [comment]
> SubscriptionIdParameter

Yes, the error is caused by `x-ms-path`. But the error is reported by `autorest.core`, not `autorest.python`. You could ask help from @timotheeguerin

**Comment by timotheeguerin** [comment]
That error does seem to be this issue that is quite old and has been fixed https://github.com/Azure/autorest/issues/3546
Not sure which job is having this issue but it is most likely using an old version of @autorest/core and should be updated. If this is is still happening with the latest please create an issue in the autorest repo.

**Comment by docohe** [comment]
@msyyc did you see @timotheeguerin comment? Is it possible to upgrade the autorest-core version in the tests?

**Comment by weidongxu-microsoft** [comment]
Java is on 3.6.6 (and the bug above is fixed in 3.0.6339)

**Comment by msyyc** [comment]
Hi @docohe @timotheeguerin 
Python is same with Java. Upgrading autorest.core does not solve the problem: https://github.com/Azure/azure-rest-api-specs/pull/17197#issuecomment-1005450932

**Comment by docohe** [comment]
@msyyc could you please open an issue to the autorest-core?

**Comment by timotheeguerin** [comment]
@docohe @msyyc found the issue. There is a typo in the spec you are using `x-ms-path` instead of `x-ms-paths` so in the Swagger-> OpenAPI 3 converter it doesn't convert `x-ms-path` to OpenAPI3 including the ref but then down the line it will try to replace all the $ref it find in the document which include that one being `#/parameters/SubscriptionId` instead of `#/components/parameters/SubscriptionId`.
If you change to the correct name you'll have some other validation error as some things are defined incorrectly in that operation.

I'll see if I can improve something so autorest can give a better error in cases like that.

**Comment by docohe** [comment]
@timotheeguerin could you please refer me to a swagger that has 'x-ms-paths' I cannot see what I'm doing wrong. I've looked in this repo and couldn't find anything. 
Do I need to use different parameters in the x-ms-paths? Not sure I understand the issue.

**Comment by timotheeguerin** [comment]
@docohe what is the current error you are having, `x-ms-paths` is the same as `paths` it just gets merged together.
Samples https://github.com/Azure/autorest.testserver/blob/0779a7f2a6a92b76be4bd8e6c57dd72897400323/swagger/paging.json#L535

**Comment by docohe** [comment]
@timotheeguerin @msyyc maybe one of you could help me. When I am using the following path in the x-ms-paths /subscriptions/{subscriptionId}/providers/Microsoft.Kusto/locations/{location}/operationresults/{operationId}?operationResultResponseType={operationResultResponseType} 

The semantic validation fails because it thinks that there is another path parameter named operationResultResponseType which is not defined. However, operationResultResponseType is a query parameter. Do you know how to help me with?

When I use the following path:
/subscriptions/{subscriptionId}/providers/Microsoft.Kusto/locations/{location}/operationresults/{operationId}
and mention the query parameter in the parameters list, it says that it has a conflict.

**Comment by timotheeguerin** [comment]
@docohe can we step back and understand what you are trying to achieve? What is this new API supposed to return?

 I am actually not sure you can use `x-ms-paths` like that with a query param that is not a fixed value. If it worked  you might need to make the parameter a `path` param not a query param then.

**Comment by docohe** [comment]
@weidongxu-microsoft we are almost ready for merge. 
We have four test that are still failing:

1. Api readiness - we are deploying our manifest with the new Feb API-version. The rollout of the manifest is about week, so this should be done by next week.
2. Cross-Version Breaking Changes - I've added the tag 'BreakingChangeReviewRequried' did someone review this? Can you tag the person who should approve this? We add a default value to a property, it had default value in the backend from day one, so there is no really breaking change here. We just forgot to add it in the swagger and it was caught by S360 
3. ~[Staging] Cross-Version Breaking Changes - is the same is 2. Please correct me if I'm wrong
4. Swagger SDK Breaking Change Tracking - I'm not familiar with this test and I'm not sure what to do with. The error message tells me nothing.

We really would like to be able to release this API version next week.

**Comment by weidongxu-microsoft** [comment]
@docohe 
1. "api readyness" You can leave it there. If urgent, we can merge the PR before service rollout. If not urgent, we can re-run this after rollout.
2. "corss-version breaking changes" It seems the only change is the default value of one property (that is missing from previous version). You may ping JeffreyRichter for a quick approve.
3. "sdk breaking changes" @ArcturusZhang for double check on Go SDK breaking change.
4. You can ignore the [staging] CI

**Comment by weidongxu-microsoft** [comment]
@docohe

Do you have new API (there is too many commits in this PR)? If yes, you also need ARM review.

**Comment by docohe** [comment]
@weidongxu-microsoft yes there is new api-version in the PR. 
There are a lot of files in this PR, but most of them changed because of the json prettify, you can tell the diff to ignore space changes.

In addition, I'll get the breaking change approval

File: specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2022-02-01/kusto.json
Comment (weidongxu-microsoft) [inline specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2022-02-01/kusto.json:187]: If it is `x-ms-long-running-operation`, I don't think you need to explicitly specify `Azure-AsyncOperation` as it is well-known header.

File: specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2022-02-01/kusto.json
Comment (weidongxu-microsoft) [inline specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2022-02-01/kusto.json:3585]: seems to be new API

**Review by weidongxu-microsoft — COMMENTED** [review]
I've looped in ARM review.

**Comment by mentat9** [comment]
@docohe, couple of things are needed before your PR is ready for ARM review:

1. Please study the materials described in the submission form and attest to that by checking the box: **[ ] Ensure you've reviewed following guidelines including ARM resource provider contract and REST guidelines. Estimated time (4 hours). This is required before you can request review from ARM API Review board.**
2. Please create a new PR following this item (and check the box when you submit):
**[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.**
Ensure the files for the current API version are copied without modification to the new API version folder for the first commit, then apply changes in subsequent commits. This is required for ARM review.

**Comment by docohe** [comment]
@mentat9 I've done what I've asked on this PR using force -f. Could you please take a look?

**Comment by mentat9** [comment]
>     "/subscriptions/{subscriptionId}/providers/Microsoft.Kusto/locations/{location}/operationresults/{operationId}?operationResultResponseType=Location": {

operationResults (casing)?

---
Refers to: specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2022-02-01/kusto.json:3585 in 93d336b. [](commit_id = 93d336bb77125cc818f68e847161b51bd4c023de, deletion_comment = False)

**Comment by mentat9** [comment]
>           "default": "Disabled"

This is a cross-version breaking change. Clients on the previous version who take the default will start getting different behavior after you switch to the new API version.

I see you got a breaking change approval. Did it cover this change?

---
Refers to: specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2022-02-01/kusto.json:3775 in 93d336b. [](commit_id = 93d336bb77125cc818f68e847161b51bd4c023de, deletion_comment = False)

**Comment by mentat9** [comment]
>         "allowMultiDatabase": {

ARM recommends string enums over boolean. Generally enums make better properties than booleans. They are more descriptive, flexible and future-proof while being easier for customers to discover, understand and use. Ref: [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review).

Best approach is to choose a better property (e.g. databaseRoutingPolicy: { "Single", "Multiple" } or similar). This type of property is generally easier for customers to understand and use. 

You can also have a string enum with True/False (or Enabled/Disabled, etc.) values which is still better than boolean, since modelAsString enums allow adding values later without requiring a new API version (e.g. "Unknown", "Default", etc.).


---
Refers to: specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2022-02-01/kusto.json:4295 in 93d336b. [](commit_id = 93d336bb77125cc818f68e847161b51bd4c023de, deletion_comment = False)

**Comment by mentat9** [comment]
>         "allowMultiDatabase": {

Similar consideration here and in the EventGridConnectionProperties definition below.

---
Refers to: specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2022-02-01/kusto.json:4340 in 93d336b. [](commit_id = 93d336bb77125cc818f68e847161b51bd4c023de, deletion_comment = False)

**Comment by mentat9** [comment]
> @mentat9 I've done what I've asked on this PR using force -f. Could you please take a look?

Thank-you @docohe, much appreciated.

**Comment by mentat9** [comment]
Signed off from ARM perspective.

**Review by weidongxu-microsoft — APPROVED** [review]
LGTM, let me know when to merge.

**Comment by docohe** [comment]
@weidongxu-microsoft could you please merge?

**Pull request merged**
